persistence/sql/persister_code.go

```diff
@@ -0,0 +1,96 @@
+package sql
+
+import (
+	"context"
+	"crypto/subtle"
+	"fmt"
+	"github.com/gobuffalo/pop/v6"
+	"github.com/gofrs/uuid"
+	"github.com/ory/kratos/selfservice/strategy/code"
+	"github.com/ory/x/sqlcon"
+	"github.com/pkg/errors"
+	"time"
+)
+
+type oneTimeCodeProvider interface {
+	GetID() uuid.UUID
+	Validate() error
+	TableName(ctx context.Context) string
+	GetHMACCode() string
+}
+
+func useOneTimeCode[P any, U interface {
```

Comment:
This code was the same for login, verification, recovery, and sign up, so I moved it into a common function. Uses some generic magic to make it work nicely. Tried it with duck typing only at first, but it caused issues and would have required reflection.